### ISSUE TITLE
Build Failures on Windows: CMake Version Conflicts and Missing prompt_template.dart When Running Example App

### ISSUE THREAD

**Issue by KJ-Waller** [BrutalCoding/aub.ai#5]
Hi there! This plugin looks really cool, seems like you've done a lot of work on this. Very impressive. 

I'm trying to test out the example app on Android from my Windows 10 machine. I'm very inexperienced with flutter, but I've been able to test out previous flutter projects on my own devices before which worked fine, but when I try to run `flutter run` for the example app I get an error related to the cmake version installed.

```
flutter run
Launching lib\main.dart on Pixel 8 in debug mode...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':aub_ai:configureCMakeDebug[arm64-v8a]'.
> [CXX1405] error when building with cmake using [Project Path]\src\CMakeLists.txt: -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: [Android SDK Path]\ndk\23.1.7779620\toolchains\llvm\prebuilt\windows-x86_64\bin\clang.exe - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: [Android SDK Path]\ndk\23.1.7779620\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring incomplete, errors occurred!
  See also "[Project Path]\android\.cxx\Debug\[Build ID]\arm64-v8a\CMakeFiles\CMakeOutput.log".

  C++ build system [configure] failed while executing:
      @echo off
      "[Android SDK Path]\cmake\3.18.1\bin\cmake.exe" ^
        "-H[Project Path]\src" ^
        "-DCMAKE_SYSTEM_NAME=Android" ^
        "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON" ^
        "-DCMAKE_SYSTEM_VERSION=28" ^
        "-DANDROID_PLATFORM=android-28" ^
        "-DANDROID_ABI=arm64-v8a" ^
        "-DCMAKE_ANDROID_ARCH_ABI=arm64-v8a" ^
        "-DANDROID_NDK=[Android SDK Path]\ndk\23.1.7779620" ^
        "-DCMAKE_ANDROID_NDK=[Android SDK Path]\ndk\23.1.7779620" ^
        "-DCMAKE_TOOLCHAIN_FILE=[Android SDK Path]\ndk\23.1.7779620\build\cmake\android.toolchain.cmake" ^
        "-DCMAKE_MAKE_PROGRAM=[Android SDK Path]\cmake\3.18.1\bin\ninja.exe" ^
        "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=[Project Path]\example\build\aub_ai\intermediates\cxx\Debug\[Build ID]\obj\arm64-v8a" ^
        "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=[Project Path]\example\build\aub_ai\intermediates\cxx\Debug\[Build ID]\obj\arm64-v8a" ^
        "-DCMAKE_BUILD_TYPE=Debug" ^
        "-B[Project Path]\android\.cxx\Debug\[Build ID]\arm64-v8a" ^
        -GNinja
    from [Project Path]\android
  CMake Warning at [Android SDK Path]\ndk\23.1.7779620\build\cmake\android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    [Android SDK Path]\ndk\23.1.7779620\build\cmake\android.toolchain.cmake:55 (include)
    [Android SDK Path]\cmake\3.18.1\share\cmake-3.18\Modules\CMakeDetermineSystem.cmake:93 (include)
    CMakeLists.txt:6 (project)


  CMake Error at CMakeLists.txt:9 (add_subdirectory):
    The source directory

      [Project Path]\src\llama.cpp

    does not contain a CMakeLists.txt file.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 14s
Running Gradle task 'assembleDebug'...                             15.5s
Error: Gradle task assembleDebug failed with exit code 1
```

I've tried installing a newer version of cmake from Android Studio's SDK Manager but even when I uninstall cmake v3.18.1 and installed 3.22.1 it always tries to use 3.18. I figured it's tied to the NDK version then so I tried installing NDK 26.1 and uninstalling 23.1 but then when I run `flutter run` it automatically installs 23.1 and runs into the same error as above. 

I then tried pointing aub_ai in `pubspec.yaml` to the version of aub_ai to use rather than using the locally cloned repo `aub_ai: ^1.0.3` instead of `aub_ai \n path: ../` but this gets me to the following error, where it seems prompt_template.dart is missing
```
flutter run
Launching lib\main.dart on Pixel 8 in debug mode...
lib/main.dart:4:8: Error: Error when reading '[...]/AppData/Local/Pub/Cache/hosted/pub.dev/aub_ai-1.0.3/lib/data/prompt_template.dart': The system cannot find the path specified.

import 'package:aub_ai/data/prompt_template.dart';
       ^
lib/main.dart:41:3: Error: Type 'PromptTemplate' not found.
  PromptTemplate promptTemplate = PromptTemplate.chatML();
  ^^^^^^^^^^^^^^
lib/main.dart:41:3: Error: 'PromptTemplate' isn't a type.
  PromptTemplate promptTemplate = PromptTemplate.chatML();
  ^^^^^^^^^^^^^^
lib/main.dart:41:35: Error: Undefined name 'PromptTemplate'.
  PromptTemplate promptTemplate = PromptTemplate.chatML();
                                  ^^^^^^^^^^^^^^
lib/main.dart:313:5: Error: 'PromptTemplate' isn't a type.
    PromptTemplate? tmpPromptTemplate;
    ^^^^^^^^^^^^^^
lib/main.dart:322:27: Error: The method 'PromptTemplate' isn't defined for the class '_MyAppState'.
 - '_MyAppState' is from 'package:aub_ai_example/main.dart' ('lib/main.dart').
Try correcting the name to the name of an existing method, or defining a method named 'PromptTemplate'.
      tmpPromptTemplate = PromptTemplate(
                          ^^^^^^^^^^^^^^
Target kernel_snapshot failed: Exception

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command '[Flutter SDK Path]\bin\flutter.bat'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 11s
Running Gradle task 'assembleDebug'...                             11.6s
Error: Gradle task assembleDebug failed with exit code 1

```

Not sure if this problem is related to my flutter setup, so here is the output of `flutter doctor -v` just in case
```
flutter doctor -v
[√] Flutter (Channel master, 3.17.0-9.0.pre.13, on Microsoft Windows [Version 10.0.19045.3803], locale en-NL)
    • Flutter version 3.17.0-9.0.pre.13 on channel master at C:\Users\KJW\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dfa694ec8f (5 weeks ago), 2023-11-14 16:10:17 -0500
    • Engine revision f15b259fe9
    • Dart version 3.3.0 (build 3.3.0-127.0.dev)
    • DevTools version 2.30.0-dev.4

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\KJW\AppData\Local\Android\Sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = C:\Users\KJW\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-10027231)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Build Tools 2019 16.11.30)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
    • Visual Studio Build Tools 2019 version 16.11.34031.81
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2022.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-10027231)

[√] VS Code (version 1.85.0)
    • VS Code at C:\Users\KJW\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.78.0

[√] Connected device (4 available)
    • Pixel 8 (mobile)  • 38031FDJH00MTA • android-arm64  • Android 14 (API 34)
    • Windows (desktop) • windows        • windows-x64    • Microsoft Windows [Version 10.0.19045.3803]
    • Chrome (web)      • chrome         • web-javascript • Google Chrome 119.0.6045.200
    • Edge (web)        • edge           • web-javascript • Microsoft Edge 120.0.2210.77

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

Thanks for this plugin btw, I really hope I can get this working and can start playing around with it.

Groetjes uit het moederland 🇳🇱

**Comment by BrutalCoding** [comment]
Hey goos!

I'm flying back tomorrow to my other home country: Aussie land 🇦🇺. But, I'll try my best to solve this before my flight.

You're based in NL, right? I'm happy to pair up over a Google Meet and figure this out together.  If my guesses didn't work for you, send me a DM anywhere and we'll arrange a time. I think it'll be solved within 30mins.

Alrighty, let me start with an estimated guess based on the good amount of details you shared:

----

I see that the error is due to a missing file: "CMakeLists.txt" seems to be missing in this folder: "aub.ai" > "src" > "llama.cpp" > (missing file)

I'm thinking that the whole llama.cpp folder is empty, because it's a git submodule which is like, hmm, an extra repository that requires a special git command to get fetched/downloaded. You need to run "git submodule update --remote". Read below.

Detailed steps:
1. Open command line (cmd), or use the one built-in VSCode for example.
2. Go to the root folder of this repository "aub.ai". For example: "cd C:\my_favorite_projects\aub.ai\"
3. Run "git submodule update --remote" (assuming you have "git" already installed)
4. Verify that there are a bunch of files now in aub.ai > src > llama.cpp (this folder should no longer be empty)

Looking good? Great. Try running the example app again in with flutter run (in the example folder).

--- 

Not looking good? Let's try another solution:

1. Open the program "git bash", it will look like a black boring terminal. This allows us to run bash scripts/commands like on Linux and macOS machines.

2. Navigate to this folder: "aub.ai" > "brutalcoding_scripts"

3. There should be a bash script that I pushed very recently named "autoupdate_windows.sh". I highly recommend to check what the file does, read the code. Never trust a random script. But assuming you did check it, run it like this:
"./autoupdate_windows.sh". 

Done. Try running the example app again as described earlier.

If there's a file execution permission error, try doing this: "chmod +x autoupdate_windows.sh". That will give the script permission to be executed.


---

It's 2am and I'm typing this on my phone on top of my head, be aware for any typo's in the commands or filenames I've mentioned. But I think that these 2 solutions would solve your issue.

Off to bed now, slaapze 💤!

**Comment by BrutalCoding** [comment]
Almost forgot: thanks for mentioning the prompt_template file issue with the pub.dev version. I wasn't aware of this. 

I'll try to reproduce that issue and also push an update manually, somewhere today or otherwise before this weekend.

The version on pub.dev is slightly lacking behind this repo. Haven't finished my CI/CD/Fastlane stuff yet to keep both repo and pub synced. 

Normally, you wouldn't need to do any of the steps I mentioned earlier with the pub.dev version. That one is plug and play. That is, if the pub.dev worked of course, my bad. Fix is coming.

**Comment by KJ-Waller** [comment]
Thanks for the detailed reply! Really appreciated. Hope you had a good flight!

I followed your instructions. After running `git submodule update --remote` from the project root, `./src/llama.cpp` was still empty, so I checked out the `./brutalcoding_scripts/autoupdate_windows.sh` script and ran it using git bash. 

```
aub.ai/brutalcoding_scripts (main)
$ ./autoupdate_windows.sh 
[AUB.AI] Your system information:
[AUB.AI] MINGW64_NT-10.0-19045 DESKTOP-VCT31J6 3.0.7-338.x86_64 2019-04-30 21:52 UTC x86_64 Msys
[AUB.AI] Navigating to root directory of the project...
[AUB.AI] Updating git submodules...
Resolving dependencies... (1.8s)
+ _fe_analyzer_shared 65.0.0
+ analyzer 6.3.0
+ analyzer_plugin 0.11.3  
+ args 2.4.2
+ async 2.11.0
+ boolean_selector 2.1.1  
+ build 2.4.1
+ build_config 1.1.1      
+ build_daemon 4.0.1      
+ build_resolvers 2.4.2   
+ build_runner 2.4.7      
+ build_runner_core 7.2.11
+ built_collection 5.1.1  
+ built_value 8.8.1       
+ characters 1.3.0        
+ checked_yaml 2.0.3      
+ ci 0.1.0
+ cli_util 0.4.1
+ clock 1.1.1
+ code_builder 4.9.0      
+ collection 1.18.0       
+ convert 3.1.1
+ crypto 3.0.3
+ custom_lint 0.5.7
+ custom_lint_builder 0.5.7
+ custom_lint_core 0.5.7
+ dart_style 2.3.4
+ fake_async 1.3.1
+ ffi 2.1.0
+ ffigen 10.0.0
+ file 7.0.0
+ fixnum 1.1.0
+ flutter 0.0.0 from sdk flutter
+ flutter_lints 3.0.1
+ flutter_test 0.0.0 from sdk flutter
+ freezed 2.4.6
+ freezed_annotation 2.4.1
+ frontend_server_client 3.2.0
+ glob 2.1.2
+ graphs 2.3.1
+ hotreloader 4.1.0
+ http_multi_server 3.2.1
+ http_parser 4.0.2
+ io 1.0.4
+ js 0.6.7
+ json_annotation 4.8.1
+ json_serializable 6.7.1
+ leak_tracker 9.0.16 (9.0.17 available)
+ leak_tracker_testing 1.0.5 (1.0.6 available)
+ lints 3.0.0
+ logging 1.2.0
+ matcher 0.12.16 (0.12.16+1 available)
+ material_color_utilities 0.8.0
+ meta 1.11.0
+ mime 1.0.4
+ package_config 2.1.0
+ path 1.8.3 (1.9.0 available)
+ plugin_platform_interface 2.1.7
+ pool 1.5.1
+ pub_semver 2.1.4
+ pubspec_parse 1.2.3
+ quiver 3.2.1
+ riverpod 2.4.9
+ riverpod_analyzer_utils 0.5.0
+ riverpod_annotation 2.3.3
+ riverpod_generator 2.3.9
+ riverpod_lint 2.3.7
+ rxdart 0.27.7
+ shelf 1.4.1
+ shelf_web_socket 1.0.4
+ sky_engine 0.0.99 from sdk flutter
+ source_gen 1.5.0
+ source_helper 1.3.4
+ source_span 1.10.0
+ sprintf 7.0.0
+ stack_trace 1.11.1
+ state_notifier 1.0.0
+ stream_channel 2.1.2
+ stream_transform 2.1.0
+ string_scanner 1.2.0
+ term_glyph 1.2.1
+ test_api 0.6.1 (0.7.0 available)
+ timing 1.0.1
+ typed_data 1.3.2
+ uuid 4.3.1
+ vector_math 2.1.4
+ vm_service 13.0.0 (14.0.0 available)
+ watcher 1.1.0
+ web 0.3.0 (0.4.0 available)
+ web_socket_channel 2.4.0
+ yaml 3.1.2
+ yaml_edit 2.1.1
Changed 92 dependencies!
7 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
Resolving dependencies in .\example... 
Got dependencies in .\example.
Building package executable... (5.6s)
Built build_runner:build_runner.
[INFO] Generating build script completed, took 315ms
[INFO] Precompiling build script... completed, took 6.1s
[INFO] Building new asset graph completed, took 804ms
[INFO] Checking for unexpected pre-existing outputs. completed, took 2ms
[INFO] Generating SDK summary completed, took 3.7s
[INFO] Running build completed, took 5.0s
[INFO] Caching finalized dependency graph completed, took 59ms
[INFO] Succeeded after 5.1s with 10 outputs (29 actions)
Deleting .dart_tool...                                               3ms
Deleting Generated.xcconfig...                                       0ms
Deleting flutter_export_environment.sh...                            0ms
Deleting ephemeral...                                                1ms
Deleting ephemeral...                                                0ms
Deleting ephemeral...                                                1ms
Deleting .flutter-plugins-dependencies...                            0ms
Deleting .flutter-plugins...                                         0ms
Resolving dependencies... 
  intl 0.18.1 (0.19.0 available)
  leak_tracker 9.0.16 (9.0.17 available)
  leak_tracker_testing 1.0.5 (1.0.6 available)
  matcher 0.12.16 (0.12.16+1 available)
  path 1.8.3 (1.9.0 available)
  test_api 0.6.1 (0.7.0 available)
  vm_service 13.0.0 (14.0.0 available)
  web 0.3.0 (0.4.0 available)
Got dependencies!
8 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
[AUB.AI] Generating a native libraries for the C++ code...
CMake Warning:
  Ignoring extra path from command line:

   ".."


CMake Error: The source directory "[PROJECT PATH]/aub.ai/src/llama.cpp" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
Error: could not load cache
[AUB.AI] Navigating back to the root directory...
[AUB.AI] Adding llama.cpp to the Windows project...
cp: cannot stat 'src/llama.cpp/build/bin/Release/llama.dll': No such file or directory
```

It looks like it updated some dependencies, but from the error it looks like CMakeLists.txt and llama.dll are still missing in `./aub.ai/src/llama.cpp/`. After running the `autoupdate_windows.sh`, this is what my `src` directy looks like:
```
$ tree src/
src/
├── CMakeLists.txt
├── aub_ai.h
├── aub_ai.mm
├── ios-cmake
└── llama.cpp
    └── build
```

Previous to this command, llama.cpp was empty, and after running `autoupdate_windows.sh`, it looks like it only created a build directory.

Thanks a lot for your help btw. Groetjes!

**Comment by BrutalCoding** [comment]
Just got off a meeting with @KJ-Waller to see if I could help solve the issue preventing him from debugging the example app. 

Long story short: Issue resolved. The `aub.ai/src/llama.cpp` folder was empty and this was resolved by git cloning the [llama.cpp](https://github.com/ggerganov/llama.cpp) repository manually.

**Here's what happened in a somewhat FAQ kinda troubleshooting style:**

**Q:** The folder aub.ai > src > llama.cpp was still empty. It should be full of files from the llama.cpp project, but it wasn't. Why is that?
**A:** This is because that folder, "llama.cpp", is not part of this repo directly. It is something special: a so called 'git submodule'. Executing 'git pull' or a git clone on aub.ai won't pull the files of this empty "llama.cpp" folder automatically. 

**Q:** How do I populate this empty "llama.cpp" folder?
**A:** After cloning or pulling this repo (aub.ai), execute this:
```git submodule update --remote```. Running this command should print/output a bunch of files that just got pulled/downloaded. Done, it's fixed. Now try running the example app again with `flutter run`. 

**However, for some unknown reason this `git submodule update --remote` command didn't work as expected for @KJ-Waller**. I haven't figured out the root cause of the issue yet, but we solved it anyways by manually adding all the files of the empty "llama.cpp" folder described below.

**Alternative fix** 
1. Download or git clone [llama.cpp](https://github.com/ggerganov/llama.cpp) in this directory (aub.ai > src > llama.cpp > ...missing files...`).
2. Verify that this "llama.cpp" folder is no longer empty.
3. Navigate to the example folder of aub.ai.
4. Execute `flutter run` again. It should work this time around.

Here's the original llama.cpp repository:
https://github.com/ggerganov/llama.cpp
